README.md
# STOAT Introduction

## Summary

The Spatiotemporal Observation Annotation Tool (STOAT) is a platform for the annotation of spatiotemporal biodiversity data (e.g. occurrence records) with values from remote sensing and other gridded environmental datasets. Environmental annotation is conducted by retrieving layer values associated with the spatiotemporal coordinates of biodiversity data. STOAT's cloud-based computations simplify and automate the environmental annotation process, eliminating the need for users to download or interface directly with complex environmental layers. STOAT brings versatile environmental annotations to the fingertips of a broader range of scientists, and contributes to analysis workflows for rapidly accumulating biodiversity and environmental data.

STOAT is hosted by Map of Life (https://mol.org) as part of its broader biodiversity informatics platform. Please see the STOAT project homepage at (https://mol.org/stoat).

rstoat is the R package interface to STOAT. rstoat allows for the easy integration of STOAT into existing R-based workflows, and provides all the core functionalities of the STOAT web application, most importantly the submission of annotation tasks, and the retrieval of annotation results. The public repository for rstoat can be found at: https://github.com/MapofLife/rstoat.

## Background

Advances in remote sensing combined with rapidly growing types and amounts of spatiotemporal biodiversity data now enable an unrivaled opportunity for planetary-scale monitoring of biodiversity change. This project builds on earlier developments of global remote-sensing-supported climate and environmental layers for biodiversity assessment, and represents a general workflow that allows for the environmental annotation, visualization, and change assessment for past and future spatiotemporal biodiversity occurrence data.

Observed, in-situ biodiversity data have intrinsic spatiotemporal grains and associated uncertainties based on observation methodology and data collection. Likewise, remotely-sensed environmental data also vary in spatiotemporal grain from meters to kilometers. This tool has the technical infrastructure and software workflows to develop and serve appropriate summaries of environmental data for biodiversity observations. For example, a list of migrating birds observed from one location one afternoon would require a different summary of environmental data than a list of vascular plants known to exist in a 100km2 protected area. This system draws from a near-real-time collection of RS and RS-derived environmental data (such as land surface temperature, precipitation, and vegetation indices) to enable both historical and near-real-time annotation of continuously updating biodiversity data streams.

The generalized software workflows and tools enable characterization and comparison of environmental associations of individuals, populations or species over time, globally. This will allow the quantification of observed environmental niches as well as the detection of change through time in both the environmental associations and geographic distributions of biodiversity.

## Environmental Data

The environmental data available through this tool come from a variety of sources, including remote sensing products (e.g. MODIS), and derived products created for a diversity of ecological purposes. Products are broken into categories depending on their temporal grain. We define dynamic products as those with finer-than-monthly temporal grain. Annual products are those with a layer per year (currently there are no layers in STOAT with a Monthly temporal grain, though that designation may be used in the future). Finally, static products are those with a single layer across all time scales. Layers provided on STOAT launch include:

*Layer, Spatial Grain, Temporal Grain*
* MODIS NDVI/EVI, 250m, 1-day
* MODIS LST Day/Night, 1km, 1-day
* Landsat 7 EVI, 30m, 16-day
* Landsat 8 EVI, 30m, 16-day
* CHELSA/EarthEnv Daily precipitation, 1km, 1-day
* ESA CCI land cover, 300m, Annual
* SRTM variables, 30m, Static
* TNC Global Human Modification, 1000m, Static
* MODIS Winter/Summer EVI, 1000m, Static
* EarthEnv layers (various), 1000m, Static

For a full and up-to-date list of layers available in STOAT, visit https://mol.org/stoat/sources.

### Processing

All derived environmental datasets, unless specified, are unmodified from their original sources. Environmental layers are hosted by [Descartes Labs](https://www.descarteslabs.com), an environmental data repository and refinery, as well as Google Earth Engine, and are accessed through a series of internal API calls. Documentation for all included datasets can be found at (https://mol.org/stoat/sources).

Remote sensing products such as MODIS and Landsat are filtered to keep only high quality observations (removing clouds and other poor quality data). Specific metadata for each product (including the quality control filtering) are provided with downloaded data.

## Biodiversity Data

Using the batch annotator, occurrence data should be first uploaded into an individual's Map of Life account before commencing annotation. Data for upload take the format of a CSV with columns for the species scientific name, latitude, longitude, and date (See https://mol.org/upload/ for more details). Data can come from a variety of sources. Online repositories such as GBIF provide huge quantities of species data free of charge. More frequently, a user may wish to annotate a private or personal dataset. Data uploaded can be marked as private to restrict access to only the data owner.

## Spatial and Temporal Buffers

STOAT provides customizable spatial and temporal buffers around annotated points. These buffers allow for users to control the effective grain size of their biodiversity data, allowing users to accommodate differences in data grain, data uncertainty, and the extents of studied ecological processes. Due to increased computational demands associated with larger buffer sizes, there are hard limits on the maximum buffer sizes: these limits vary by product and scale with the spatial and temporal grain of a product. In R, buffer limits can be viewed with the function get_products(); in the web application they represent the ends of the buffer slider. Tasks with buffer requests beyond limits will be rejected by the annotator. The units of spatial and temporal buffers are meters, and days, respectively. Static layers should be requested with a temporal buffer of 0 days.

## Attribution

STOAT was developed by the Map of Life team at Yale University in collaboration with researchers at the University of Florida and the University at Buffalo. The STOAT team can be contacted at (https://mol.org/contact-us) with any inquiries.

STOAT currently has a manuscript in review. More citation info will follow.

## Funding

Funding for the development of STOAT is provided by NASA grants AIST-16-0092 and AIST-18-0034, as well as NSF grant DEB-1441737.

# STOAT User Guide

For a full demonstration of the package's functionalities as well as example code, please see the Introduction vignette provided.

For answers to Frequently Asked Questions, please refer to the [STOAT website](https://mol.org/stoat/faq).

## Installation
To install rstoat, please run the following line of code (requires remotes package).

```r
remotes::install_github('mapoflife/rstoat', ref='main')
```

## Usage

To use STOAT, one must first [create a Map of Life account](https://auth.mol.org/register). Once an account has been created, a user can upload datasets to be annotated through the [Map of Life uploader](https://mol.org/upload/datasets). *At the current time, data upload must be carried out from the Map of Life website.* All further steps of the annotation process may be carried out either on the web application or in this R package. A list of functions is provided below, most notably "start_annotation_batch()"; see the Introduction vignette for example code.

Additionally, a simple annotation function "start_annotation_simple()"" is provided to allow for limited annotation capacity (1000 records) without the need to use the web interface of the Map of Life Uploader. This function has a significantly lower record limit, nor does it serve the complete set of layers provided by the full annotator, but finds use as a testing tool and convenient means of running a pilot annotation. The simple annotator is distinct in its code from the full annotator, utilizing Google Earth Engine rather than Descartes Labs as a data source. Slight discrepancies in annotations are attributable to this, though any severe discrepancies should be reported to the STOAT team. We strongly recommend use of the full annotator for research applications.

### Package Functions:

Login to your Map of Life account

```r
mol_login("your.email@address"")
```

List your Map of Life datasets:

```r
my_datasets()
```

Start a custom annotation job request: 

```r
start_annotation_batch("<dataset_id>", "My annotation task", layers = c("product-layer-sbuff-tbuff", "product-layer-sbuff-tbuff"))
```

Start a simple annotation request - these do not require use of the uploader, but have a lower limit of occurrences, for use for testing or with pilot data: 
```r
start_annotation_simple("<events_dataframe>", layers = c("product-layer-sbuff-tbuff", "product-layer-sbuff-tbuff"))
```

Get a list of environmental products available for annotation, along with limits to spatial and temporal buffers:

```r
get_products()
```

List your past and current annotation jobs:

```r
my_jobs()
```

Get details on a specific custom annotation job: 

```r
job_details("<annotation_id>")
```

Get a list of species that were annotated in a custom annotation request: 

```r
job_species("<annotation_id>")
```

Download completed results of a custom annotation request: 

```r
download_annotation("<annotation_id>")
```

Read and unpack the downloaded data into a single data.frame:

```r
read_output("path/to/your/extracted/annotation")
```

Download example data (used in vignette)

```r
download_example_data("destination/path")
```



NEWS.md
# rstoat 1.0.3

* Updated to move keyring package into suggests from imports.

# rstoat 1.0.2

* Updated start_annotate_simple documentation to reflect backend updates.

# rstoat 1.0.1

* Added a `NEWS.md` file to track changes to the package.
* Removed unused data imports.
* Moved vignette data into the package, instead of relying on remote connection to fetch data.


R/annotation.R

extract_layer <- function(layers) {
  if (list %in% is(layers)) {
    layers <- lapply(layers, function (slayer) {
      layer = strsplit(slayer, "-")[[1]]
      if (length(layer) != 4) {
        stop(paste("Layer", slayer, "must be in the format '<PRODUCT CODE>-<VARIABLE>-<SPATIAL BUFFER>-<TEMPORAL BUFFER>'"))
      }
      list(
        product = layer[1], variable = layer[2], spatial = as.numeric(layer[3]), temporal = as.numeric(layer[4])
      )
    })
  }
  layers
}

# modis_lst_day_1000_1 -> modis-lst_day-1000-1
# landsat8_evi_30_16   -> landsat8-evi-30-16
internal_to_code <- function (code) {
  b = strsplit(code, '_')
  sapply(b, function(a){
    n <- length(a)
    if (n == 4) {
      return(paste0(a, collapse = '-'))
    } else {
      return(paste0(c(a[1], # product
                      paste(a[2:3], collapse = '_'), # variable
                      a[4], a[5]), collapse = '-'))
    }
  })
}

download_annotation_err <- function (e) {
  if(!curl::has_internet()) {
    message('The rstoat package requires an internet connection, please connect to the internet.')
    return(NULL)
  } else {
    message('File not found. Please check that your annotation has completed.')
    return(NULL)
  }
}

#' @title Start batch annotation
#'
#' @description Submit a dataset previously uploaded on mol.org for annotation.
#' To upload a dataset please visit https://mol.org/upload/
#' Requires login, please run mol_login(<email_address>)
#'
#' @param dataset_id The id of the dataset. List your datasets with the my_datasets() function.
#' @param title The title of the annotation job.
#' @param layers A list of parameters or vector of codes, of the layers, see the examples below.
#'
#' @return No return value, check my_jobs() to confirm successful job submission.
#' @export
#'
#' @examples
#' \dontrun{
#' start_annotation_batch('<dataset_id>', 'My annotation task', layers = list(
#'    list(product = "chelsa", variable = "precip", spatial = 1000, temporal = 30)
#' ))
#' # alternatively supplying the code is fine.
#' start_annotation_batch('<datset_id>',
#'   'My 2nd annotation task', layers = c("modis-ndvi-1000-1", "modis-lst_day-1000-1"))
#' }
start_annotation_batch <- function (dataset_id, title, layers) {
  body <- list(
    dataset_id = dataset_id,
    title = title,
    layers = extract_layer(layers)
  )
  resp <- post_json('user','annotate','dataset', body = body)
  message(paste(resp$status, resp$message))
}

#' @title Start simple annotation
#'
#' @description Submit a dataframe for on-the-fly annotation.
#' Does not require login - for use for small numbers of records and pilot jobs.
#'
#' You can annotate using layers that are on earthengine!

#' Layer parameters:
#' Used by all:
#' \itemize{
#' \item spatial: The spatial buffer in meters.
#' \item temporal: The temporal buffer in days.
#' }
#'
#' For STOAT layers:
#' \itemize{
#'   \item product: The product e.g "srtm", or "landsat8".
#'   \item variable; The vairable, e.g "elevation", or "evi".
#' }
#'
#' For non STOAT layers, ie those in Google Earth Engine:
#' \itemize{
#' \item id: The id of the image in Google Earth Engine, mapped to "product" in the output.
#' \item static: Whether to load the imagery as an ImageCollection or as an Image.
#' \item bands: A list, wiht one element, which is used to specify which band of the imagery to use for the annotation, mapped to "variable" in the output.
#' \item reducers: A list containing one or more of the following: \itemize{
#'    \item mean
#'    \item lcv_count
#'    \item mode
#'    \item median
#'    \item stdev
#'    \item min
#'    \item max
#'    \item stdev
#'  } If only one is provided the output will be assigned to "value", else the output will be named the same as the reducer.
#'
#' }
#'
#'
#' @param events A data.frame for on the fly annotation
#' @param layers A list of parameters or vector of codes, of the layers, see the examples below.
#' @param coords A vector of length 2 containing column names for record longitudes, and latitudes.
#' @param date Column name for record dates, dates must take the format YYYY-MM-DD
#'
#'
#' @return Input data.frame with values from the annotation appended, in addition to unique identifier field event_id.
#' \itemize{
#'  \item event_id: A unique identifier for each occurrence
#'  \item product: Product used for annotation
#'  \item variable: Variable used for annotation
#'  \item s_buff: Spatial buffer in meters applied to occurrence
#'  \item t_buff: Temporal buffer in days applied to occurrence
#'  \item value: Annotated value of occurrence from requested layer (mean within buffer),
#'               if there is only one reducer (default), then this value will be here.
#'  \item stdev: Standard deviation of values within buffer
#'  \item valid_pixel_count: Number of pixels within buffered area'
#' }
#' @export
#'
#' @examples
#' \dontrun{
#' events <- data.frame(
#'    event_id = as.character(1:2),
#'    lng = c(-4, 24),
#'    lat = c(10, 10),
#'    date = '2015-01-01'
#' )
#'
#' # simple layer string format: PRODUCT-VARIABLE-S_BUFF-T_BUFF
#' layers <- 'landsat8-evi-100-16'
#' start_annotation_simple(events, layers)
#'
#' start_annotation_simple(events, layers)
#'
#' # For lcv_count (Count of landcover value), 'value' returned is a string of
#' # landcover counts within the AOI.
#' # The output format is:
#' #  <LANDCOVER_CLASS>:<COUNT_OF_PIXELS_WITH_THAT_CLASS>
#' #  classes are seperated by commas.
#'
#' start_annotation_simple(events, list(
#'   list(
#'     id="COPERNICUS/Landcover/100m/Proba-V-C3/Global",
#'     s_buff=1000,
#'     reducers=list("lcv_count", "mode"),
#'     static=FALSE,
#'     t_buff=365,
#'     bands=list("discrete_classification")
#'   )
#' ))
#'
#' # Annotating with two worldclim layers:
#' #   bio01 is annual mean temperature
#' #   bio12 is annual precipitation
#'
#' start_annotation_simple(events, list(
#'  list(
#'    "id"= "WORLDCLIM/V1/BIO",
#'    "s_buff"=1000,
#'    "reducers"=list("mean"),
#'    "static"= TRUE,
#'    "t_buff"= 1,
#'    "bands"=list("bio01")
#'  ),
#'  list(
#'    "id"= "WORLDCLIM/V1/BIO",
#'    "s_buff"=1000,
#'    "reducers"=list("mean"),
#'    "static"= TRUE,
#'    "t_buff"= 1,
#'    "bands"=list("bio12")
#'  )
#' ))

#' }
start_annotation_simple <- function (events, layers, coords=c('lng','lat'), date='date') {
  if (any(!(c(coords[1], coords[2], date) %in% names(events)))) {
    stop("Dataframe must contain either columns 'lng', 'lat', 'date', or user must provide arguments for alternate column names")}
  events_subset <- data.frame(event_id = events$event_id,
                              lng = events[[coords[1]]],
                              lat = events[[coords[2]]],
                              date = events[[date]],
                              stringsAsFactors = F)
  body <- list(
    events = events_subset,
    params = extract_layer(layers)
  )
  resp <- post_json('annotate/ondemand', body = body, otf=T, authenticate=FALSE)

  resp <- merge(events, resp, by = 'event_id')
  resp
}

#' @title Download annotation results
#'
#' @description Download results of a successfully completed batch annotation.
#' Requires login, please run mol_login(<email_address>)
#' Uses the output from my_jobs() for the annotation id.
#'
#' @param annotation_id The id of the annotation
#' @param dir The directory where to write the annotation.
#' @return The path of the downloaded annotation.
#' @export
#'
#' @examples
#' \dontrun{
#' download_annotation(<annotation_id>, <dir>)
#' }
download_annotation <- function (annotation_id, dir = 'annotation_results') {
  url <- build_url('user/annotations', annotation_id, 'download')
  annotation_url <- get_resp({
    httr::GET(url, get_auth_header(), ua())
  })
  if (is.null(annotation_url)) return(NULL)
  annotation_url <- annotation_url$url
  tmp_path <- tempfile()
  download_path <- paste0(dir, '/', annotation_id)
  if (!dir.exists(dir)) {
    dir.create(dir)
    message(paste0('Created directory: ', dir))
  }
  if (dir.exists(download_path)) {
    message('Annotation already downloaded.')
    return(NULL)
  }

  status <- NULL
  # catch file not found error
  tryCatch ({
    status <- utils::download.file(annotation_url, tmp_path, mode='wb')
  }, error = download_annotation_err, warning = download_annotation_err
  )
  if (!is.null(status)) {
    if (status == 0){
      message('Unzipping')
      utils::unzip(tmp_path, exdir=download_path)
      message(paste0('Annotation available at: ', download_path))
      return(download_path)
    } else {
      message('There was an error downloading your annotation')
      return(NULL)
    }
  }
  else {
    message('There was an error downloading your annotation')
  }
}

#' @title Read annotation output into R
#' @description Convenience function which reads and joins annotation results spread across multiple files for space efficiency.
#' To run this function, please first download annotated data first using download_annotation()
#'
#' @param directory The path of the data.
#' @param drop_event_id Whether to drop the event_id column or not.
#'
#' @return A data.frame of annotated data, one row per variable per event
#' @export
#'
#' @examples
#' \dontrun{
#' read_output("path/to/your/downloaded/data/directory")
#' }
read_output <- function (directory, drop_event_id = TRUE) {
  files <- list.files(directory, full.names = T)
  species_file <- files[grepl('species.csv$', files)]
  event_file <- files[grepl('events.csv$', files)]
  events <- utils::read.csv(event_file)
  events$event_id <- as.character(events$event_id)
  species <- utils::read.csv(species_file)
  species$event_id <- as.character(species$event_id)

  if (!requireNamespace("dplyr", quietly = TRUE)) {
    print("Merging using base R. Install package \"dplyr\" for faster merges")
    out <- merge(species, events, by="event_id")
  }
  else {
    out <- dplyr::left_join(species, events, by="event_id")
  }

  output_files <- files[grepl('results.csv$', files)]

  for (output_file in output_files) {
    current_file <- utils::read.csv(output_file)
    current_file$event_id <- as.character(current_file$event_id)
    current_name <- basename(output_file)
    current_name <- sub("_results.csv", "", current_name)
    # create a new column in 'out' named after each output file
    current_file[current_name] <- current_file$value
    current_file <- current_file[, c("event_id", current_name)]

    if (!requireNamespace("dplyr", quietly = TRUE)) {
      out <- merge(out, current_file, by="event_id")
    }
    else {
      out <- dplyr::left_join(out, current_file, by="event_id")
    }
  }

  if (drop_event_id) out$event_id <- NULL
  out
}
